feat: add support for externalized consumers (#7657)
Kong/docs.konghq.com: app/konnect/centralized-consumers.md

---
title: Centralized Consumers

> alpha: true

Reviewer: ```suggestion
```
This is going to GA. See https://kongstrong.slack.com/archives/C06FJ632KJS/p1741905058701429?thread_ts=1741903435.247969&cid=C06FJ632KJS

---

A [Consumer](/gateway/latest/key-concepts/consumers/) is a {{site.base_gateway}} entity that consumes or uses the APIs managed by {{site.base_gateway}}. These can be things like applications, services, or users who interact with your APIs. Consumers can be scoped to a {{site.konnect_short_name}} region and managed centrally (centralized Consumers) or be scoped to a control plane in Gateway Manager.

Centralized Consumers provide the following benefits:

> * The Consumer identity can be setup centrally instead of being defined across multiple control planes.

Reviewer: ```suggestion
* The Consumer identity can be set up centrally instead of being defined across multiple control planes.
```

* Share Consumers across multiple control planes. Users don't need to replicate changes to Consumer identity in multiple control planes and Consumer configuration doesn't conflict.

> * Reduces configuration sync issues between the control plane and the dataplane. Centralized Consumers aren't part of the configuration that is pushed down from the control plane to the dataplane, so it reduces config size and latency. 

Reviewer: ```suggestion
* Reduces configuration sync issues between the control plane and the data planes. Centralized Consumers aren't part of the configuration that is pushed down from the control plane to the data planes, so it reduces config size and latency. 
```

## How Centralized Consumers work

When you create a centralized Consumer, you must assign it to a realm. A realm groups centralized Consumers around an identity, defined by organizational boundaries, such as a production realm or a development realm. Realms are connected to a [geographic region](/konnect/geo/) in {{site.konnect_short_name}}.

The following diagram shows how you can configure centralized Consumers in realms as well as scope Consumers to a Control Plane:

> <!--vale off -->

Reviewer: I don't really understand what this diagram is showing me. 
For example:

> Consumer: Ira
> Consumer Groups: Gold for appointments

Does this consumer only have access to the appointments control plane? Or does it have access to both control planes, but since the consumer group Gold only exists on `appointments`, it can only access appointments? Or something else?

Are control planes located inside realms, or alongside them?

{% mermaid %}
flowchart TD

> A(Consumer: Cruz<br>Consumer Groups: Gold for appointments, Silver for billing)

Reviewer 2: Each consumer can be in multiple consumer groups, but you can't scope the consumer group to a service.

If `billing` respected the `Gold` group, then Cruz would be Gold for billing too.

B(Consumer: Ira<br>Consumer Groups: Gold for appointments)
C(Route: /appointments)
D(Gold tier: 1000 req/60 sec)
E(Route: /dev)
F(Consumer: Izumi)
G(Consumer: Sasha)
H(Route: /billing)
I(Silver tier: 10 req/60 sec)

subgraph id1 [Prod realm]
direction LR
    A
    B
end
subgraph id2 [Dev realm]
direction LR
    G
end

id1 --> id3
id1 --> id4
id2 --> id5

subgraph id3 [Appointments Control Plane]
direction LR
C
D
end

subgraph id4 [Billing Control Plane]
direction LR
H
I
end

subgraph id5 [Dev Control Plane]
direction LR
F
E
end

style id1 stroke-dasharray:3,rx:10,ry:10
style id2 stroke-dasharray:3,rx:10,ry:10
style id3 stroke-dasharray:3,rx:10,ry:10
style id4 stroke-dasharray:3,rx:10,ry:10
style id5 stroke-dasharray:3,rx:10,ry:10
{% endmermaid %}
<!-- vale on-->

## Create centralized Consumers and Realms 

You can create centralized Consumers using the {{site.konnect_short_name}} API. Only Org Admins have CRUD permissions for centralized Consumers. 

> 1. Use the following custom {{site.base_gateway}} image when you create a [Data Plane node](/konnect/gateway-manager/data-plane-nodes/):
>    ```
>    kong/kong-gateway-dev
>    ```

Reviewer: ```suggestion
{:.important}
> Centralized Consumers are available for {{site.base_gateway}} 3.10 data planes. Make sure your control plane uses 3.10 data planes at minimum when configuring centralized Consumers.

```
Something like this, and feel free to rephrase.
This is an internal docker repo. As of next week, when 3.10 goes live, the 3.10 image will have externalized consumers support, so they just need to make sure to use 3.10 data planes.

1. Use the `/realms` endpoint to create a realm and optionally associate it with allowed Control Planes and time-to-live values:
   ```
   curl -X POST \
   https://{region}.api.konghq.com/v1/realms \
   -H "Content-Type: application/json" \
   -H "Authorization: Bearer $KONNECT_TOKEN" \
   -d '{
        "name": "prod",
        "allowed_control_planes": [
            "$CONTROL_PLANE_UUID"
        ],
        "ttl": 10,
        "negative_ttl": 10,
        "consumer_groups": [
            "$CONSUMER_GROUP"
        ]
    }'
   ```
   Save the ID of the realm.

   Be sure to replace the following with your own values:
   * {region}: Region for your {{site.konnect_short_name}} instance.
   * $KONNECT_TOKEN: Replace with your {{site.konnect_short_name}} personal access token.
   * $CONTROL_PLANE_UUID: (Optional) Replace with your Control Plane UUID.

>    * `ttl` and `negative_ttl`: (Optional) Time-to-live in minutes of the consumer or negative consumer for this realm in the {{site.base_gateway}} cache.

Reviewer 2: What's a `negative consumer`? I don't think that this is what `negative_ttl` means. @zekth can you help us out here?

Reviewer 3: there's no concept of `negative_consumer`. `negative_ttl` represents the TTL of a bad login cache entry. Meaning api key `foobar` is sent to the gateway; gateway reaches the service to authenticate it; if it fails the api key `foobar` will be labelled as non working in gateway cache to the time of the TTL. During that time it won't try to reache the authentication server.

   * $CONSUMER_GROUP: (Optional) Replace with the name of the consumer groups you want to associate with the realm.
1. Use the `/realms/{realmId}/consumers` endpoint to create a Consumer and optionally assign it to a Consumer Group:
   ```
   curl -X POST \
   https://{region}.api.konghq.com/v1/realms/{realmId}/consumers \
   -H "Content-Type: application/json" \
   -H "Authorization: Bearer TOKEN" \
   -d '{
         "username": "$CONSUMER_NAME",
         "consumer_groups": ["$CONSUMER_GROUP"]
       }'
   ```
   Be sure to replace the following with your own values:
   * {region}: Region for your {{site.konnect_short_name}} instance.
   * $KONNECT_TOKEN: Replace with your {{site.konnect_short_name}} personal access token.
   * {realmId}: The ID of the realm you created previously. 
   * $CONSUMER_NAME: Replace with the name of the consumer.

>    * $CONSUMER_GROUP: (Optional) Replace with the name of the consumer groups you want to associate with the consumer.

Reviewer 2: If I specify `consumer_groups` on a consumer does that replace `consumer_groups` from the realm, or are they added together?

If I specify `consumer_groups: []` does that mean no consumer groups, or use the groups from the realm?

Reviewer 4: This is additive here, I'll add that info to the doc.

1. Consumers require authentication. Configure authentication using the [Key Auth plugin](/hub/kong-inc/key-auth/#configure-realms-for-centralized-consumers-in-konnect).

{:.note}

> > **Note:** If you are using KIC to manage your Data Plane nodes in {{site.konnect_short_name}}, ensure that you configure the `telemetry_endpoint` in the Data Plane. You can find the `telemetry_endpoint` in the {{site.konnect_short_name}} UI in [Gateway Manager]() in the Data Plane node instructions.

Reviewer: ```suggestion
> **Note:** If you are using KIC to manage your Data Plane nodes in {{site.konnect_short_name}}, ensure that you configure the `telemetry_endpoint` in the Data Plane. You can find the `telemetry_endpoint` in the {{site.konnect_short_name}} UI in [Gateway Manager](https://cloud.konghq.com/gateway-manager/) in the Data Plane node instructions.
```
